### PR TITLE
fix the problem that Service node port e2e test may failed if node ip not reachable

### LINKED ISSUE #57819: Service node port e2e test may failed if node ip not reachable
**Is this a BUG REPORT or FEATURE REQUEST?**:

> /kind feature

**What happened**:

in nodeport test case, if the nodeIP is not reachable for same reason(firewall, network isolate...), then the `http://nodeIP:nodeport` definitely can't be access, in this case, the e2e test will result in failure, that maybe a little Arbitrary.
For example, in kubeadm builded gce cluster, all node use the internal IP, which is not reachable from the outside,so the nodeport test will fail even if nodeport type service have been created and work well.

**What you expected to happen**:

The nodeport test don't relay on cluster node reachable or not from outside the cluster.

**Environment**:
- Kubernetes version (use `kubectl version`): v1.10.0-alpha.1.311+da9a4d5da9ecc6
- Cloud provider or hardware configuration: gce
- OS (e.g. from /etc/os-release): ubuntu-1604-xenial-v20160420c
- Kernel (e.g. `uname -a`):
- Install tools: kubeadm
- Others:

### PR DESCRIPTION
**Pull request by Lion-Wei**
**What this PR does / why we need it**:
Access the node port inside the cluster when test nodeport type service.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #57819

**Special notes for your reviewer**:

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Comment by m1093782566** [comment]
@Lion-Wei 

Can you explain why skip the test instead of return error?

File: test/e2e/network/service.go [hunk 1]
```diff
@@ -458,6 +458,9 @@ var _ = SIGDescribe("Services", func() {
 
 		jig := framework.NewServiceTestJig(cs, serviceName)
 		nodeIP := framework.PickNodeIP(jig.Client) // for later
+		if re, _ := framework.TestNotReachableHTTP(nodeIP, 10250); re {
+			framework.Skipf("Service node port test needs node ip is reachable, skip.")
```
Comment (m1093782566) [inline test/e2e/network/service.go]: why skip instead of returning error?

**Comment by Lion-Wei** [comment]
@m1093782566 THanks for replaying.
The issue is, in nodeport test case, if the nodeIP is not reachable for same reason(firewall, network isolate...), then the `http://nodeIP:nodeport` definitely can't be access, in this case, the e2e test will result in failure, that maybe a little Arbitrary.
For example, in kubeadm builded gce cluster, all node use the internal IP, which is not reachable from the outside,so the nodeport test will fail even if nodeport type service have been created and work well.

My solution is access the nodeport service inside the cluster node, to avoid nodeIP unreachable problem. 
@MrHohn , Can you help me take a look? thanks.

File: test/e2e/network/service.go [hunk 2]
```diff
@@ -57,10 +57,10 @@ var _ = SIGDescribe("Services", func() {
 		if CurrentGinkgoTestDescription().Failed {
 			framework.DescribeSvc(f.Namespace.Name)
 		}
-		for _, lb := range serviceLBNames {
-			framework.Logf("cleaning gce resource for %s", lb)
-			framework.CleanupServiceGCEResources(cs, lb, framework.TestContext.CloudConfig.Zone)
-		}
+		//for _, lb := range serviceLBNames {
```
Comment (m1093782566) [inline test/e2e/network/service.go]: remove dead code.

Comment (m1093782566) [reply test/e2e/network/service.go]: And, why remove these lines?

**Comment by m1093782566** [comment]
My question is:

Is the test case running? Why it didn't cause a problem? And, nodeport can be visible outside cluster.

Per associated issue, "For example, in kubeadm builded gce cluster, all node use the internal IP, which is not reachable from the outside, so the nodeport test will fail even if nodeport type service have been created and work well." 

The issue is caused by kubeadm?

File: test/e2e/network/service.go (hunk 2 above)
Comment (Lion-Wei) [reply test/e2e/network/service.go]: oops, that's a mistake, I did this for local test.

**Comment by m1093782566** [comment]
I feel like adding test for internal node IP is good, however, we should test external node IP as well. Is there anyway we can fix from kubeadm side?

**Comment by Lion-Wei** [comment]
@m1093782566 GCE build cluster all node have both external IP and internal IP, like: 
```
NAME                             ZONE           MACHINE_TYPE   PREEMPTIBLE  INTERNAL_IP  EXTERNAL_IP     STATUS
bootstrap-e2e-master             us-central1-f  n1-standard-1               10.128.0.2   35.192.166.147  RUNNING
bootstrap-e2e-minion-group-38k1  us-central1-f  n1-standard-2               10.128.0.5   35.202.157.180  RUNNING
bootstrap-e2e-minion-group-7v74  us-central1-f  n1-standard-2               10.128.0.6   35.225.121.25   RUNNING
bootstrap-e2e-minion-group-91rr  us-central1-f  n1-standard-2               10.128.0.4   35.226.249.159  RUNNING
bootstrap-e2e-minion-group-r4b8  us-central1-f  n1-standard-2               10.128.0.3   35.226.227.19   RUNNING
```
And in this test case, will access the externalip+nodeport, which is reachable:
```
I0105 02:08:39.993] STEP: hitting the pod through the service's NodePort
I0105 02:08:39.993] Jan  5 02:08:21.334: INFO: Testing HTTP reachability of http://35.202.157.180:31505/echo?msg=hello
I0105 02:08:39.993] Jan  5 02:08:21.336: INFO: Got error testing for reachability of http://35.202.157.180:31505/echo?msg=hello: Get http://35.202.157.180:31505/echo?msg=hello: dial tcp 35.202.157.180:31505: getsockopt: connection refused
I0105 02:08:39.996] Jan  5 02:08:29.336: INFO: Testing HTTP reachability of http://35.202.157.180:31505/echo?msg=hello
```
But in kubeadm build cluster, only have internal IP, which can't be reachable from outside.
```
STEP: hitting the pod through the service's NodePort
Jan  4 12:02:38.460: INFO: Testing HTTP reachability of http://10.128.0.5:30024/echo?msg=hello
Jan  4 12:02:38.462: INFO: Got error testing for reachability of http://10.128.0.5:30024/echo?msg=hello: Get http://10.128.0.5:30024/echo?msg=hello: dial tcp 10.128.0.5:30024: getsockopt: connection refused
Jan  4 12:02:40.462: INFO: Testing HTTP reachability of http://10.128.0.5:30024/echo?msg=hello
Jan  4 12:02:40.463: INFO: Got error testing for reachability of http://10.128.0.5:30024/echo?msg=hello: Get http://10.128.0.5:30024/echo?msg=hello: dial tcp 10.128.0.5:30024: getsockopt: connection refused
Jan  4 12:02:42.462: INFO: Testing HTTP reachability of http://10.128.0.5:30024/echo?msg=hello
Jan  4 12:02:42.463: INFO: Got error testing for reachability of http://10.128.0.5:30024/echo?msg=hello: Get http://10.128.0.5:30024/echo?msg=hello: dial tcp 10.128.0.5:30024: getsockopt: connection refused
```

Nodeport can be visible outside cluster, but in this case, our problem is the host is not reachable,. : (

**Comment by Lion-Wei** [comment]
In ipvs CI job, we used kubernetes-anywhere and kubeadm build cluster, I'm not sure which part have problem, but I'll try to find out. @m1093782566

**Comment by danwinship** [comment]
I think it's assumed/required that all nodes are accessible to the test suite. Eg, this test calls `framework.PickNodeIP()`, which calls `framework.GetNodePublicIps()`, which does:

    ips := CollectAddresses(nodes, v1.NodeExternalIP)
    if len(ips) == 0 {
    	// If ExternalIP isn't set, assume the test programs can reach the InternalIP
    	ips = CollectAddresses(nodes, v1.NodeInternalIP)
    }

(That comment comes from #49025; previously there was an equivalent but implicit assumption involving LegacyHostIP.)

**Comment by Lion-Wei** [comment]
@danwinship Thanks for your information. I notice this comment too, but this assumption not always correct, and in this case, may result in the test case result incorrect. : (

**Comment by danwinship** [comment]
What I was trying to say was, I think the tests are *expected* to fail if the host running the tests can't reach all the nodes. Like, basically, this is a bug in the setup of the test environment, not a problem with the tests.

Or at least, if it must be possible to run the tests in environments where nodes aren't reachable, then those tests should be tagged in some way so that they can be skipped by people running them in such environments, rather than having the test itself decide what to do (since if the node is *unintentionally* unreachable, you would want the test to fail).

I'm not sure what the right way to tag it would be though... you can't use "[Feature:...]" because then it would be skipped by default. But there doesn't seem to be a standard tag syntax for "things that should be tested by default but which some people might want to skip"...

**Comment by errordeveloper** [comment]
I agree with @danwinship.

**Comment by Lion-Wei** [comment]
I think this pr can be closed, since it's not appropriate anymore.

**Pull request closed without merging**